src/docker/dockerUtils.ts

```diff
@@ -0,0 +1,110 @@
+    const cwd = shellOpts.cwd || vscode.workspace.rootPath;
+    const image = await getDefaultImageName(cwd, imagePrefix);
+    await buildDockerImage(dockerClient, image, shellOpts);
+    if (imagePrefix) {
+        await pushDockerImage(dockerClient, image, shellOpts);
+    }
+    return image;
+}
+
+function sanitiseTag(name : string) {
+    // Name components may contain lowercase letters, digits and separators.
+    // A separator is defined as a period, one or two underscores, or one or
+    // more dashes. A name component may not start or end with a separator.
+    // https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+
+    return name.toLowerCase().replace(/[^a-z0-9._-]/g, '-');
+}
+
+async function getDefaultImageName(cwd: string, imagePrefix?: string): Promise<string> {
+    const name = sanitiseTag(path.basename(cwd));
+    const version = await findVersion(cwd);
+    let image = `${name}:${version}`;
+    if (imagePrefix) {
+        image = imagePrefix + "/" + image;
+    }
+    return image;
+}
+
+async function findVersion(cwd: string): Promise<string> {
+    const shellOpts = Object.assign({ }, shell.execOpts(), { cwd });
+    const shellResult = await shell.execCore('git describe --always --dirty', shellOpts);
+    return shellResult.code !== 0 ? "latest" : shellResult.stdout.trim();
+}
+
+async function buildDockerImage(dockerClient: DockerClient, image: string, shellOpts: any): Promise<void> {
+    // Build docker image.
+    const buildResult = await shell.execCore(`${dockerClient} build -t ${image} .`, shellOpts);
+    if (buildResult.code !== 0) {
+        throw new Error(`Image build failed: ${buildResult.stderr}`);
+    }
+    kubeChannel.showOutput(image + ' built.');
+}
+
+async function pushDockerImage(dockerClient: string, image: string, shellOpts: any): Promise<void> {
+    // Push docker image.
+    const pushResult = await shell.execCore(`${dockerClient} push ${image}`, shellOpts);
+    if (pushResult.code !== 0) {
+        throw new Error(`Image push failed: ${pushResult.stderr}`);
+    }
+    kubeChannel.showOutput(image + ' pushed.');
+}
+
+/**
+ * When using the command "minikube docker-env" to get the local kubernetes docker env, it needs run with the admin privilege.
+ * To workaround this, this function will try to resolve the equivalent docker env from kubeconfig instead.
+ */
+export async function resolveDockerEnv(kubectl: Kubectl): Promise<{}> {
+    const dockerEnv = {};
+    dockerEnv["DOCKER_API_VERSION"] = await dockerApiVersion();
+    const currentCluster = await getCurrentClusterConfig(kubectl);
+    if (!currentCluster || !currentCluster.server || !currentCluster.certificateAuthority) {
+        return {};
+    }
+    
+    if (/^https/.test(currentCluster.server)) {
+        dockerEnv["DOCKER_TLS_VERIFY"] = 1;
+    }
+    const serverUrl = url.parse(currentCluster.server);
+    dockerEnv["DOCKER_HOST"] = `tcp://${serverUrl.hostname}:2376`;

```

Comment:
If I'm pointing at a cluster in the cloud (rather than a local minikube), what is going to happen here?  Do we need to trap for this situation?

Reply:
Actually for cloud, such as `aks`, this method will return an emtpy `{}` because its certificate-authority in kubeconfig is empty. It just has value in local minikube.